lib/cext/preprocess.rb

```diff
@@ -95,8 +95,168 @@
+                            }
+
+NOKOGIRI_TUPLE_DOC_POINTER_PATCH = { match: "tuple->doc = rb_doc",
+                              replacement: "tuple->doc = NEW_WEAK_REF(rb_doc)"
+                            }
+NOKOGIRI_TUPLE_CACHE_POINTER_PATCH = { match: "tuple->node_cache = cache",
+                              replacement: "tuple->node_cache = NEW_WEAK_REF(cache)"
+                            }
+
+NOKOGIRI_CONTEXT_CAST_PATCH = { match: "(VALUE)ctx;",
+                              replacement: "GET_LOCAL_REF(ctx);"
+                            }
+NOKOGIRI_CONTEXT_CAST_PATCH_2 = { match: "(VALUE)ctx;",
+                              replacement: "GET_GLOBAL_REF(ctx);"
+                            }
+NOKOGIRI_CONTEXT_USERDATA_CAST_PATCH = { match: "(VALUE)(ctx->context->userData);",
+                              replacement: "GET_LOCAL_REF(ctx->context->userData);"
+                            }
+NOKOGIRI_DATA_CAST_PATCH = { match: "(VALUE)data;",
+                              replacement: "GET_LOCAL_REF(data);"
+                            }
+
+NOKOGIRI_HANDLER_CAST_PATCH = { match: "(VALUE)xsltGetExtData(transform, functionURI);",
+                              replacement: "GET_GLOBAL_REF(xsltGetExtData(transform, functionURI));"
+                            }
+
+NOKOGIRI_NODE_PRIVATE_CAST_PATCH = { match: "(VALUE)node->_private;",
+                              replacement: "GET_WEAK_REF(node->_private);"
+                            }
+
+NOKOGIRI_CTX_STYLE_CAST_PATCH = { match: "ctxt->style->_private",
+                              replacement: "GET_WEAK_REF(ctxt->style->_private)"
+                            }
+
+NOKOGIRI_FUNC_INSTANCE_CAST_PATCH = { match: "(wrapper->func_instances",
+                              replacement: "(GET_GLOBAL_REF(wrapper->func_instances)"
+                                    }
+
+NOKOGIRI_XPATH_ERROR_PATCH_1 = { match: "VALUE thing = Qnil;",
+                              replacement: "VALUE thing = Qnil;\n  VALUE errors = rb_ary_new();"
+                                    }
+
+NOKOGIRI_XPATH_ERROR_PATCH_2 = { match: "xmlSetStructuredErrorFunc(NULL, Nokogiri_error_raise);",
+                              replacement: "xmlSetStructuredErrorFunc(NEW_LOCAL_REF(errors), Nokogiri_error_array_pusher);"
+                                    }
+NOKOGIRI_XPATH_ERROR_PATCH_3 = { match: "if(xpath == NULL)",
+                              replacement: "if (RARRAY_LEN(errors) > 0) { rb_exc_raise(rb_ary_entry(errors, 0)); }\nif(xpath == NULL)"
+                               }
+NOKOGIRI_WRAP_ERROR_PATCH_1 = { match: 'VALUE msg',
+                                replacement: 'VALUE msg[1]'
+                              }
+NOKOGIRI_WRAP_ERROR_PATCH_2 = { match: 'msg = ',
+                                replacement: 'msg[0] = '
+                              }
+NOKOGIRI_WRAP_ERROR_PATCH_3 = { match: '&msg',
+                                replacement: 'msg'
+                              }
```

Comment:
If these constants are only used in one place below, could you inline them?
It makes reading them significantly harder.

Reply:
The `NO_ASSIGNMENT` constant should remain last, next to `read_field`.